pakt.1
.TH "PAKT" "1" "October 2024" "pakt v0.1.0" "User Commands"

.SH NAME
pakt \- categorize installed Arch Linux packages

.SH SYNOPSIS
\fIpakt\fR <pacman or AUR helper arguments + categories>

.SH DESCRIPTION
pakt is a POSIX-compliant shell wrapper around pacman or an AUR helper of your choice. It adds "categories" as a new feature of your package manager. It means that while you install packages, you can assign tags in order to group them and gain an overview over your system.

The package names will be put into files with the name of the respective categories. The path to look for is either \fI$XDG_DATA_HOME/pakt/\fR or \fI~/.local/share/pakt/\fR.

.SH OPTIONS
You should be able to alias pakt to the name of your package manager, therefore the arguments are the same, with the exception of category flags.

+\fIcategory1\fR
.br
       Assigns the packages written before to \fIcategory1\fR.

+\fIcategory1\fR,\fIcategory2\fR,\fIcategory3\fR
.br
       Assigns the packages written before to \fIcategory1\fR, \fIcategory2\fR and \fIcategory3\fR.

++
.br
       Assigns the packages written before to all currently existing categories.

If you include these flags when uninstalling packages, their names will be removed from the respective categories. Otherwise, they will stay in the category files.

.SH EXAMPLES
\fIpakt\fR -S neovim
.br
       Categories are optional, of course.

\fIpakt\fR -S neovim +cli
.br
       Category syntax in its simplest form.

\fIpakt\fR -S neovim +cli pipewire +sound
.br
       neovim gets assigned to \fBcli\fR, pipewire to \fBsound\fR.
.br
       You can write as much of these pairs as you like.

\fIpakt\fR -S neovim +core,cli,dev htop
.br
       Categories follow a comma-separated syntax.
.br
       neovim gets assigned to \fBcore\fR, \fBcli\fR and \fBdev\fR, htop to none.

\fIpakt\fR -S neovim ++
.br
       neovim gets assigned to all categories that have a file
.br
       at \fI$XDG_DATA_HOME/pakt/\fR or \fI~/.local/share/pakt/\fR.

\fIpakt\fR -Rcnsu neovim +dev
.br
       neovim will be uninstalled and removed from the \fBdev\fR category,
.br
       not from \fBcore\fR and \fBcli\fR though.

\fIpakt\fR -Rcnsu neovim ++
.br
       neovim will be uninstalled and removed from all its categories,
.br
       including \fBcore\fR, \fBcli\fR and \fBdev\fR.

.SH ENVIRONMENT
PAKT_CONF_PATH
.br
       Path of the \fIpakt.conf\fR configuration file. Defaults to \fI/etc\fR.

.SH FILES
\fI/etc/pakt.conf\fR
.br
       Default address of the configuration file. Look into the file for more information.

\fI/etc/pakt.conf.default\fR
.br
       Default configuration options for \fIpakt.conf\fR.
.br
       It's also a part of the souce code at \fIpakt.conf.default\fR.

\fI~/.local/share/pakt/\fR
.br
       Path containing category files, if \fI$XDG_DATA_HOME\fR is not set.
.br
       Category files are plain lists of packages assigned to the respective category.

.SH AUTHORS
Sergey Lavrent <https://github.com/hiimsergey>
.LP
MrMineDe <https://github.com/mrminede>

.SH LICENSE
GPLv2: GNU General Public License version 2
.LP
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by the law.

.SH BUGS
https://github.com/mrminede/pakt/issues

.SH SEE ALSO
\fBpakt-sync\fR(1) \fBpakt-dry\fR(1)


pakt.sh
#!/bin/sh
# TODO END CONSIDER changing pakt.sh to pakt

# maybe there is a better approach, but this allows for easy use in if statement without [] aka test command, which results in problems
stringNotContain() { case $1 in *$2* ) return 1;; *) return 0;; esac ;}

# TODO Change -h, as it is used by pacman!
if [ "$1" = "-h" ]; then
	echo "This will be the help page"
	exit
fi

# Source the config file
# TODO DECIDE if there is no config, we should copy the default config as configfile and load that instead of defining it here
# Define the location where we search for the config file
ACTIVATE_DEF_CONF=0
if [ -z "$PAKT_CONFIG_PATH" ]; then
	if [ -z "$XDG_CONFIG_HOME" ]; then
		CONF_PATH="$XDG_CONFIG_HOME/pakt/pakt.conf"
	else
		CONF_PATH="$HOME/.config/pakt/pakt.conf"
	fi
else
	CONF_PATH="$PAKT_CONFIG_PATH"
fi
# Now check if config file is there and load it
if [ -e "$CONF_PATH" ]; then
	. "$CONF_PATH"
else
	ACTIVATE_DEF_CONF=1
fi

if [ $ACTIVATE_DEF_CONF -eq 1 ]; then
	CAT="default"
	CMD="pacman"
	if [ -z "$XDG_DATA_HOME" ]; then
		PTH="$HOME/.local/share/pakt"
	else
		PTH="$XDG_DATA_HOME/pakt"
	fi
	MODE=""
fi

PAC=""
SYNC=0

# This is used, to save the state of -C argument, to store the ARG after -C
DASH_C_TRUE=0
for A in "$@"; do
	if [ "$DASH_C_TRUE" -eq 1 ]; then
		CAT="${CAT} ${A}"
		DASH_C_TRUE=0
		continue;
	fi
	case "$A" in
		-*)
			case "$A" in
				*S*)
					MODE="S"
					ARG="${ARG} ${A}" #${!i}"
				;;
				*R*)
					MODE="R"
					ARG="${ARG} ${A}"
				;;
				# TODO Add -Q capability
				"-C")
					# We want to store the argument after -C, so we set a variable that will store the next element and then skip to the next word
					DASH_C_TRUE=1
				;;
				# TODO change this to smth, that resembles sync, but is not taken by pacman
				"-ABC")
					SYNC=1
				;;
				*)
					ARG="${ARG} ${A}"
				;;
			esac
		;;
		*)
			# TODO is there anything that is not started with - but still no package?
			PAC="${PAC} ${A}"
		;;
	esac
	i="$((i+1))"
done
# If -C is the last argument, throw an error!
if [ "$DASH_C_TRUE" -eq 1 ]; then
	echo "Syntax: -C ARG ; -C needs one argument! See pakt -h for more info"
	exit
fi

# Execute pacman
if [ $SYNC -eq 0 ]; then
	if ! sudo $CMD $ARG $PAC; then
		echo "Pacman returned a fail. Exiting..."
		exit
	fi
else # -s, sync the packages
	PAC_DEL=$(sudo $CMD -Qeq)
	PAC_BASE=$(sudo $CMD -Qq)
	PAC_ADD=""
	# We want to combine all the categoriyfiles we got as parameter
	for c in $CAT; do
		# We unfortunately have to do this for every entry in every categoryfile...
		while read -r p; do
			# PAC_DEL should include all packages implicitly installed on the system by the user that are not in the categoryfiles
			PAC_DEL=$(echo "$PAC_DEL" | tr ' ' '\n' | grep -v "^$p\$")
			# PAC_ADD should include all packages of the categoryfiles, that are not installed on the system
			if stringNotContain "$(echo "$PAC_BASE" | tr '\n' ' ')" " $p "; then
				PAC_ADD="${PAC_ADD} ${p}"
			fi
		done < "$PTH/$c"
	done
	# We convert spaces to new lines for grep a few lines ago, for pacman we need to reverse that
	PAC_DEL=$(echo "$PAC_DEL" | tr '\n' ' ')
	# We want to install/remove all the packages specified in the command as well
	# TODO This does not work as intended atm
	if [ -n "$PAC" ] && [ "$MODE" = "S" ]; then
		PAC_ADD="${PAC_ADD} ${PAC}"
	fi
	if [ -n "$PAC" ] && [ "$MODE" = "R" ]; then
		PAC_DEL="${PAC_DEL} ${PAC}"
	fi
	# If there are any packages that would get installed and uninstalled, we need to throw an error
	# TODO DECIDE if we really need to throw an error or rather handle it differently
	for p in $PAC_DEL; do
		if ! stringNotContain "$(echo "$PAC_ADD" | tr '\n' ' ')" " $p "; then
			  echo "Invalid arguments! You would install and remove $p! Exiting..."
			  exit
		fi
	done
	# To be safe, we print the Packages to remove and install before calling pacman
	if [ -n "$PAC_DEL" ]; then
		echo "Removing(dependencies not listed): $PAC_DEL"
	fi
	if [ -n "$PAC_ADD" ]; then
		echo "Installing: $PAC_ADD"
	fi
	# Finally sync the packages
	# TODO Decide if sync should also do a full system upgrade.
	# This would also reduce the danger of partial upgrades/corrupt packages etc.
	if [ -n "$PAC_DEL" ]; then
		if ! sudo pacman -Rs $PAC_DEL; then
			echo "Pacman returned a fail. Exiting..."
			exit
		fi
	fi
	if [ -n "$PAC_ADD" ]; then
		if ! sudo pacman -Sy $PAC_ADD; then
			echo "Pacman returned a fail. Exiting..."
			exit
		fi
	fi
fi

# Create path if it doesnt exist
mkdir -p "$PTH"
for f in $CAT; do
	# This will make sure, that there are no error messages about a categoryfile not existing
	# These errors are not fatal, but the user should not see them!
	touch "$PTH/$f"
	for p in $PAC; do
		case $MODE in
			S) # Add Package
				# Check if Package is already in file, then dont add it
				if [ "$p" != "$(grep "^$p\$" "$PTH/$f")" ]; then
					echo "$p" >> "$PTH/$f"
				fi
				;;
			R) # Remove Package
				sed -n "/^$p\$/"'!'"p" "$PTH/$f" | tee "$PTH/$f" > /dev/null # You cant pipe directly into PTH/f bc the pipe is opened first, therefor the file is beeing cleared, therefor sed opens a empty file
				;;
		esac
	done
done
